### PR TITLE
Move detection of cutouts in Android engine to `onApplyWindowInsets`

### PR DESCRIPTION
**Pull request by yaakovschectman**
Stop including the translation of window insets cutouts to Flutter cutout DisplayFeatures in the window info listener method, which would not necessarily receive changes in the correct order/state when they update, e.g. when the screen orientation changes.

Fixes https://github.com/flutter/flutter/issues/155658

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java [hunk 1]
```diff
@@ -538,7 +539,7 @@ protected void onDetachedFromWindow() {
   @TargetApi(API_LEVELS.API_28)
   protected void setWindowInfoListenerDisplayFeatures(WindowLayoutInfo layoutInfo) {
```
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java:540]: I dont understand why only the display cutout features need to be moved and not all the features in this method. 

Unrelated to your change it is weird that this had a targetApi of 28 and still did an api check inside the method.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java [hunk 2]
```diff
@@ -649,6 +651,49 @@ public void systemInsetDisplayCutoutSimple() {
     assertEquals(100, viewportMetricsCaptor.getValue().viewInsetTop);
   }
 
+  @SuppressWarnings("deprecation")
+  @Test
+  @Config(sdk = 28)
```
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:656]: can this be minSdk?

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java (hunk 1 above)
Comment (gmackall) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java:540]: My understanding is that this listener is triggered (only) on changes to [`DisplayFeature`](https://developer.android.com/reference/kotlin/androidx/window/layout/DisplayFeature), which include hinges and display folds but (from @yaakovschectman and my manual testing) not camera cutouts (or other top of device sensors).

Those are instead reported as [`WindowInsets`](https://developer.android.com/reference/android/view/WindowInsets#getDisplayCutout()) with a nonzero element reported by `getDisplayCutout()`.

Flutter was conflating both of these two distinct groups into a `List<FlutterRenderer.DisplayFeature>`, but the list would only get updated in the case of changes to the former group, because underlying updates to the latter didn't trigger the listener.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java (hunk 2 above)
Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java]: Yes, albeit with some extra modifications

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java [hunk 3]
```diff
@@ -649,6 +651,69 @@ public void systemInsetDisplayCutoutSimple() {
     assertEquals(100, viewportMetricsCaptor.getValue().viewInsetTop);
   }
```
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: Based on the issue https://github.com/flutter/flutter/issues/155658 I think there are more regression tests we should include. 
From my reading this test is verifying that cutouts when triggered how we expect show up in viewpoint metrics. 

After Reading the bug i am thinking we should have a rotation test that verifies that cutouts are not "left" after the hardware has changed orientation. Additionally there is logic in this pr to combine displayCutouts and displayFeatures, I would expect a test to modify one and assert the other was preserved and vice versa *and* a test to show that if both are modfied the joint list is updated correctly. This second (of 3) change might be possible without adding a new test but by adding asserts to existing tests or it might require a new test depending on the existing coverage.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java [hunk 4]
```diff
@@ -649,6 +651,69 @@ public void systemInsetDisplayCutoutSimple() {
...
+      when(displayCutout.getWaterfallInsets()).thenReturn(unusedInsets);
+    }
+    when(displayCutout.getSafeInsetTop()).thenReturn(100);
+    when(displayCutout.getSafeInsetLeft()).thenReturn(100);
+    when(displayCutout.getSafeInsetBottom()).thenReturn(100);
+    when(displayCutout.getSafeInsetRight()).thenReturn(100);
+
+    flutterView.onApplyWindowInsets(windowInsets);
+
+    verify(flutterRenderer, times(2)).setViewportMetrics(viewportMetricsCaptor.capture());
```
Comment (gmackall) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:706]: Am I reading this correctly that we are updating the viewport metrics (`flutterRenderer.setViewportMetrics()`) twice when insets get applied? `sendViewportMetricsToFlutter()` only calls it once, right - do you know why we are seeing two invocations?

sorry if I'm misreading the test here, just want to be sure

**Review by gmackall — COMMENTED** [review]
mostly lgtm just one question about the tests

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java [hunk 5]
```diff
@@ -649,6 +651,69 @@ public void systemInsetDisplayCutoutSimple() {
...
+    }
+    // WindowInsets::getInsets was added in API 30.
+    if (Build.VERSION.SDK_INT >= 30) {
+      when(windowInsets.getInsets(anyInt())).thenReturn(unusedInsets);
+    }
+    // DisplayCutout::getWaterfallInsets was added in API 30.
+    if (Build.VERSION.SDK_INT >= 30) {
+      when(displayCutout.getWaterfallInsets()).thenReturn(unusedInsets);
+    }
+    when(displayCutout.getSafeInsetTop()).thenReturn(100);
```
[the remark this answers is not in the document]
Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:699]: Those values are not being tested in this test; they are checked in the above test `systemInsetDisplayCutoutSimple`. I think I can apply this feedback to that test, though

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java (hunk 4 above)
Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:706]: This is because it was already called once earlier in the test. I think there is a `clearInvocations` function I can call to avoid this; I'll have to try it

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java (hunk 3 above)
Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: For testing the rotation as you mention, this sounds like it would require an integration test to run on an emulator; I do not think we can unit-test it, as changing the orientation in a unit test will have no effect on the value returned by the stubbed methods like we are using here. Do you agree with that? And if so, are there existing integration tests in this repo that perform anything similar?

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java [hunk 6]
```diff
@@ -1485,6 +1484,10 @@ private void sendViewportMetricsToFlutter() {
     viewportMetrics.devicePixelRatio = getResources().getDisplayMetrics().density;
     viewportMetrics.physicalTouchSlop = ViewConfiguration.get(getContext()).getScaledTouchSlop();
 
+    viewportMetrics.displayFeatures.clear();
```
[the remark this answers is not in the document]
Comment (gmackall) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java:1487]: I also thought about separating the two lists within `ViewportMetrics`, and wouldn't be opposed.

I'm not too concerned about being required to remember to call clear on `viewportMetrics.displayFeatures`, though, because there the pr establishes a pattern that list is exclusively handled by this helper (`sendViewportMetricsToFlutter`) which takes care of clearing.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java (hunk 4 above)
Comment (gmackall) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:706]: got it, makes sense

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java (hunk 3 above)
Comment (reidbaker) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: I am hoping we can not do an integration test but instead do a roboletric test that runs on an activity. Possibly by modifying the orientation then recreating the activity. 
See https://stackoverflow.com/a/26468296 for inspiration.

Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: We can change the orientation in a robolectric unit test, but I still fail to see how that will help us for a unit test where the display cutouts are provided through a mock with stubbed methods for returning its metrics. The closest thing I can find is that maybe we can use robolectric's ShadowDisplay (which is deprecated and planned for removal), and to access the ShadowDisplay in the mock's stubs. I have not yet tested this approach, but if that would actually work, I am still unsure if it even reflects the behavior we want to test within reason.

Comment (reidbaker) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: I go back and forth on if an integration test is appropriate here in general we get more flakes with an integration test but also this is something our users are expecting us to handle. 

I think your suggestion to use shadowDisplay is interesting but if it is depercrated then not worth the time.

Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: Even disregarding the deprecation, I don't think using shadow displays would actually test what we are looking to test. The display cutouts are already repositioned wrt orientation by the system by the time the containing insets are passed to `onApplyWindowInsets`. All we would really accomplish by using the shadow display would be to emulate some of that logic ourselves.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java:1495]: Let us separate the lists in Viewport metrics then. I agree your helper does a good job of isolating the current logic but I think future logic is where we would get an accidental regression.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java (hunk 3 above)
Comment (reidbaker) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: Let go ahead and do a integration test that validates rotation and cutout. To me it is right on the edge of what needs /justifies an integration test but rotation logic is a classic cause of android errors and I think it is likely that if we have another rotation regression around display logic we will be thankful the test already exists.

Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: @reidbaker By this, do you mean a scenario in the scenario app? And if so, how do we specify/enforce device features such as cutouts when the app is run? I have been looking through the scenario app code and CI files for the past couple days and still can't identify what provides the "physical" characteristics of the device that runs the tests.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: I've separated the two lists, introduced a new `DisplayCutout` class, as that inherently represents less data than the generic `DisplayFeature`, and modified the view and renderer tests accordingly.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java (hunk 3 above)
Comment (reidbaker) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: I think the rotation test might need to be in the framework but as to how to define physical devices I am not aware of an api that lets you specific physical traits. There are developer settings that might be able to be set via adb that can simulate notches also we have access to physical devices and I think you can force a device to rotate also via adb. 

Now that you are digging into it I think there is additional value to the test if only for the reason of talking about how we can test things like rotation.
It could be recency bias but I think the native_driver_tests that @matanlurey added could be expanded to include rotation.

Comment (matanlurey) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: There are [rotation tests](https://github.com/flutter/flutter/blob/1cc8a07ace699de12fcd31ac9e8fcfa14f1017e2/dev/integration_tests/native_driver_test/test_driver/platform_view_blue_orange_gradient_main_test.dart#L44-L61) if that's what you mean.

Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: @reidbaker do you know if we have any existing examples of forcing a test to run on a specific physical device and/or requiring a test to be run only in the context of certain adb commands?

Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: Also I want to make sure I document here before I forget that you can get a list of cutout overlays with

    adb shell cmd overlay list | grep "display\.cutout"

and then dis/enable one with

    adb shell cmd overlay <enable/disable> <cutout package name>

Comment (yaakovschectman) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: @reidbaker Actually, I have an integration test on a local branch that can test this behavior. However, as the current state of this behavior is bugged, my understanding is that this PR will need to land first with only the engine-level unit tests before we can add this integration test to the framework. Does this sound right? If so, do you have any more pending thoughts or concerns for this PR?

Comment (reidbaker) [reply shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:652]: Let me rephrase to see if I understand. You have an integration test on a branch of flutter/flutter that fails with the current behavior and passes with a custom engine running this pr. If that is the case then yes we can land this pr then your other pr. That said, let's put that second pr up for review so that reviewers can look at what is covered in each. 

TBH having the test fail against flutter/flutter and pass with this patch is exactly the type of test we want here thank you for writing that test.

**Comment by reidbaker** [comment]
@gmackall will you give this a second pass?

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java [hunk 7]
```diff
@@ -1221,6 +1223,19 @@ public void setViewportMetrics(@NonNull ViewportMetrics viewportMetrics) {
       displayFeaturesType[i] = displayFeature.type.encodedValue;
       displayFeaturesState[i] = displayFeature.state.encodedValue;
     }
+    int cutoutOffset = viewportMetrics.displayFeatures.size() * 4;
+    for (int i = 0; i < viewportMetrics.displayCutouts.size(); i++) {
+      DisplayCutout displayCutout = viewportMetrics.displayCutouts.get(i);
+      displayFeaturesBounds[cutoutOffset + 4 * i] = displayCutout.bounds.left;
+      displayFeaturesBounds[cutoutOffset + 4 * i + 1] = displayCutout.bounds.top;
+      displayFeaturesBounds[cutoutOffset + 4 * i + 2] = displayCutout.bounds.right;
+      displayFeaturesBounds[cutoutOffset + 4 * i + 3] = displayCutout.bounds.bottom;
```
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:1232]: Consider authoring a helper here since this code is repeated and any bugs in one would likely be a bug in the other.

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java [hunk 8]
```diff
@@ -1221,6 +1223,19 @@ public void setViewportMetrics(@NonNull ViewportMetrics viewportMetrics) {
...
+      // Display cutouts always have type CUTOUT and state UNKNOWN.
+      displayFeaturesType[viewportMetrics.displayFeatures.size() + i] =
+          DisplayFeatureType.CUTOUT.encodedValue;
+      displayFeaturesState[viewportMetrics.displayFeatures.size() + i] =
```
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:1236]: It seems weird to intentionally be setting unknown here. It seems to be what were were doing before but should we have added a state enum value for these?

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java [hunk 9]
```diff
@@ -1358,11 +1377,18 @@ public DisplayFeature(Rect bounds, DisplayFeatureType type, DisplayFeatureState
...
-    public DisplayFeature(Rect bounds, DisplayFeatureType type) {
+  /**
+   * Description of a cutout on the physical display.
+   *
+   * <p>A simplified analog to {@link DisplayFeature} that specifically handles display cutouts.
+   */
+  public static final class DisplayCutout {
+    public final Rect bounds;
+
+    public DisplayCutout(Rect bounds) {
```
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:1390]: should this default to unknown? (or maybe a new state value?

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java [hunk 10]
```diff
@@ -634,21 +636,98 @@ public void systemInsetDisplayCutoutSimple() {
...
+    when(displayCutout.getSafeInsetLeft()).thenReturn(110);
+    when(displayCutout.getSafeInsetTop()).thenReturn(120);
+    when(displayCutout.getSafeInsetRight()).thenReturn(130);
+    when(displayCutout.getSafeInsetBottom()).thenReturn(140);
 
     flutterView.onApplyWindowInsets(windowInsets);
 
     verify(flutterRenderer, times(2)).setViewportMetrics(viewportMetricsCaptor.capture());
-    validateViewportMetricPadding(viewportMetricsCaptor, 200, 150, 200, 150);
+    validateViewportMetricPadding(viewportMetricsCaptor, 200, 120, 250, 140);
```
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:649]: Consider adding a comment about why this is a mix of display cutout and waterfall insets so future maintainers can understand at a glance.

File: shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java [hunk 11]
```diff
@@ -1173,6 +1271,34 @@ private void mockSystemGestureInsetsIfNeed(WindowInsets windowInsets) {
     }
   }
 
+  @SuppressWarnings("deprecation")
+  private WindowInsets setupMockDisplayCutout(List<Rect> boundingRects) {
+    WindowInsets windowInsets = mock(WindowInsets.class);
+    DisplayCutout displayCutout = mock(DisplayCutout.class);
+    when(windowInsets.getDisplayCutout()).thenReturn(displayCutout);
+    when(displayCutout.getBoundingRects()).thenReturn(boundingRects);
+    // The following mocked methods are necessary to avoid a NullPointerException when calling
```
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java:1280]: Great comment

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java (hunk 8 above)
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:1236]: The current doc-comment for `UNKNOWN` states `The display feature is a cutout or this state is new and not yet known to Flutter.` The framework also currently expects cutouts to have a state of `unknown`.

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java (hunk 9 above)
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:1390]: I'm not clear on what you are asking here. Should what default to unknown?

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java (hunk 8 above)
Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:1237]: seems odd but out of scope for this pr.

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java (hunk 9 above)
Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:1391]: The Display feature class which we previously used for cutouts has a type and a state.  Should those be hardcoded in the DisplayCutout class as type and state then let the code above in setViewportMetrics refer to DisplayCutout.type and DisplayCutout.state instead of using for example DisplayFeatureState.UNKNOWN.encodedValue.

Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:1393]: I do not have any strong opinions on it, so I will change it accordingly.

**Review by gmackall — APPROVED** [review]
One larger question and a couple of small nits.

I'm fine being overruled on the larger question, so LGTM as is! But would be interested to see if you feel similarly on the question of if we should introduce the new `DisplayCutout`

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java [hunk 12]
```diff
@@ -1358,11 +1377,21 @@ public DisplayFeature(Rect bounds, DisplayFeatureType type, DisplayFeatureState
...
     }
+  }
 
-    public DisplayFeature(Rect bounds, DisplayFeatureType type) {
+  /**
+   * Description of a cutout on the physical display.
+   *
+   * <p>A simplified analog to {@link DisplayFeature} that specifically handles display cutouts.
+   */
+  public static final class DisplayCutout {
```
[the remark this answers is not in the document]
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java]: I have no opinion, so sure

File: shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java [hunk 13]
```diff
@@ -1209,18 +1216,26 @@ public void setViewportMetrics(@NonNull ViewportMetrics viewportMetrics) {
             + "Display Features: "
             + viewportMetrics.displayFeatures.size());
```
[the remark this answers is not in the document]
Comment (yaakovschectman) [reply shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java]: Done

**Comment by yaakovschectman** [comment]
@gmackall If nothing else, using two separate collections would make any future migration to separate them on the framework easier, and simplifies our tests.
